### PR TITLE
Xenial   drop autoconnect attempts

### PR DESCRIPTION
**Pull request by Pekkari**
This patch removes automatic attempt
to connect right after a failure to
any wifi available, which often triggers
a passwrod prompt for the automatically
configured open wifi networks nearby.

Signed-off-by: José Pekkarinen jose.pekkarinen@foxhound.fi

### CODE REVIEW

**Comment by Pekkari** [comment]
@Flohack74 it seems that introducing the new Jenkinsfile yields the same result, though thanks for the hint. If you come up with any other suggestion, I'm happy to help.

**Comment by UniversalSuperBox** [comment]
@Pekkari, Jenkins is set to not "trust" changes from non-members of the repository. I've updated the original Xenial branch with the new Jenkinsfile and rebased your changes on top of it. It should build now.

**Comment by Pekkari** [comment]
Very good, thanks!

**Comment by Flohack74** [comment]
@UniversalSuperBox now that this is cleaned I wonder if we should expedite it into OTA-18, its an ask from HalloWeltSysteme, and we probably need to wait for next weeks RC anyways?

**Review by UniversalSuperBox — COMMENTED** [review]
I'm not so sure about this change. While there are reports that it works, I'm not sure why or for how long. I'm also not sure how this change could affect other, correct code paths in the timeout handling.

Note the comment in `handle_auth_or_fail(...)`:

```
/* Ask for new secrets only if we've never activated this connection
 * before.  If we've connected before, don't bother the user with
 * dialogs, just retry or fail, and if we never connect the user can
 * fix the password somewhere else.
 */
```

If the connection has been connected before, `new_secrets` gets set to `false`. Otherwise, it is set to `true`. Later, if `new_secrets` is true, `NM_SECRET_AGENT_GET_SECRETS_FLAG_REQUEST_NEW` is added to the flags which are passed to `nm_act_request_get_secrets`.

If that really long flag is unset, the secret agent should just return the previous secret from the store.

Have you confirmed that all of the assumptions here hold? That `new_secrets` is false *and* the flag is unset? If so, we may have a problem in our secret agent rather than a problem in NM.

If @Flohack74 still wants to merge this, then okay. But I don't think it's the correct solution to the problem, and I expect a regression that I can't predict from reading the code.

> This patch removes automatic attempt to connect right after a failure to any wifi available, which often triggers a passwrod prompt for the automatically configured open wifi networks nearby.

Something is very wrong here. NetworkManager should not attempt to connect to any ol' wifi network in range. Is it instead attempting to connect to a network that you've connected to before? Is it in the list of recent networks?

Can you provide a log for any of the failure cases that you're working on? NetworkManager logs to `/var/log/syslog`

**Comment by Pekkari** [comment]
handle_auth_or_fail triggers a reconnection attempt with any network that has been configured(and open networks are automatically configured without user interaction). Now that code path is to handle any errors in the connection, not only
the ones that would benefit from password prompts. Effectively, anytime that code path is reached, the timestamp that is
used to detect if we ever connected is uninitialized, and triggering the password prompt, even if you already connected
to that network. In other words, it's poorly implemented. This code path is removing the attempt of connection that request
the password, and anybody is welcome to build it properly.

Re to my comment, NetworkManager attempts to recover connection to the user mercilessly, no matter how, so if there
is an open network, it will try to see if it gets the connection over there. In my case, it attempted to connect to my
neighbours open wifi, while it got a valid credentials from my closed wifi, and it asked me for a password of an open
network that doesn't make any sense. I'll see if I can recover some logs later today.

**Comment by Pekkari** [comment]
In other thoughts, I don't get the point why this state handler do more than setting the error state and propagating the error message, since, applications should decide what to do with the error, some applications would decide to prompt the password, redirect to the configuration tool, retry 10 times before bothering the user at all, etc... Why NM have such a default hard coded
doesn't seem correct to me.

**Comment by UniversalSuperBox** [comment]
I'm only asking my questions because I do not fully understand the code in question which is being changed, and I need your help to understand it. Without that understanding, I won't be able to identify whether any future issues are related to this patch or not.

-----

> if there is an open network, it will try to see if it gets the connection over there

Please check your "Previous networks" section in Settings -> Wi-Fi. Ensure that the open networks that NetworkManager tries to connect to are not there. If this is true and NM tries to connect to any Wi-Fi network in range, it is a huge security risk. If NM is just trying to connect to a different network which it previously connected to because your primary network failed, it makes more sense.

You could get networks in your "Previous networks" list by accidentally tapping them once in the network indicator, for example.

Reading through [the 1.2.2 version of `nm-device-wifi.c`](https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/1.2.2/src/devices/wifi/nm-device-wifi.c), specifically the [handle_auth_or_fail](https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/1.2.2/src/devices/wifi/nm-device-wifi.c#L2063) implementation, I do not see where it would attempt to connect to a different network.

-----

Instead, I do see an interesting behavior: If NM has tried to reconnect more than three times (whether or not `WIRELESS_SECRETS_TRIES` gets cleared is, heh, unclear), it marks the connection as failed. For any other case:

1. If `new_secrets` is `false`, a request to the NM secret agent is fired with the flag `NM_SECRET_AGENT_GET_SECRETS_FLAG_ALLOW_INTERACTION`. This tells the agent that it should look into its credential store first and *only* prompt the user for the password if the agent does not know one.
2. If `new_secrets` is `true`, the secret agent instead gets told `NM_SECRET_AGENT_GET_SECRETS_FLAG_REQUEST_NEW`. This means that the agent *must* request a new password from the user.

If this code and the secret agent are working correctly, NetworkManager should attempt to reconnect to the network three times with the same credentials *before* asking the user for a new password. I'm personally not sure what our agent is, but it's probably indicator-network. However, I would ask you to check and make sure that this behavior is working correctly. You can check this by using `dbus-monitor` to read the communication between NetworkManager and its clients, or you could add debugging prints in the code to check the values of `new_secrets` and `flags`. We're interesting in answering only one question: "Is our secret agent really being asked to prompt for the password, or is it mishandling the request?"

If the secret agent is really being told `NM_SECRET_AGENT_GET_SECRETS_FLAG_REQUEST_NEW`, then this patch is indeed one option to resolve the problem. However I'd be worried why `new_secrets` is being set right away, maybe that is what should really change.

-----

Please don't take my comments as an attack on your work. However, I'm very concerned with the overall stability and security of Ubuntu Touch, and changing code inside network-manager in our downstream fork could very well compromise one or both of those. We need to be really careful with it. Upstream designed it a certain way for a reason, and we should be careful to understand why the walls are in place before we try to take them down.

In that case, I would recommend opening a dialog with the upstream network-manager project. Hop in to #nm on Libera Chat IRC to ask about the design of this code path instead of assuming that it is wrong. See if upstream would take this patch as it is (the code hasn't changed much in the past 6 years). If not, we should probably be redesigning it too.

**Comment by Pekkari** [comment]
> Please check your "Previous networks" section in Settings -> Wi-Fi. Ensure that the open networks that NetworkManager tries to connect to are not there. If this is true and NM tries to connect to any Wi-Fi network in range, it is a huge security risk. If NM is just trying to connect to a different network which it previously connected to because your primary network failed, it makes more sense.
> 
> You could get networks in your "Previous networks" list by accidentally tapping them once in the network indicator, for example.

Currently, and fortunately, I see only my network over there right now, but it hasn't been
always the case.

> Reading through [the 1.2.2 version of `nm-device-wifi.c`](https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/1.2.2/src/devices/wifi/nm-device-wifi.c), specifically the [handle_auth_or_fail](https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/1.2.2/src/devices/wifi/nm-device-wifi.c#L2063) implementation, I do not see where it would attempt to connect to a different network.

It attempts to connect to the connection pointed by req, which doesn't need to be the
network that failed to connect.

> 
> Instead, I do see an interesting behavior: If NM has tried to reconnect more than three times (whether or not `WIRELESS_SECRETS_TRIES` gets cleared is, heh, unclear), it marks the connection as failed. For any other case:
> 
> 1. If `new_secrets` is `false`, a request to the NM secret agent is fired with the flag `NM_SECRET_AGENT_GET_SECRETS_FLAG_ALLOW_INTERACTION`. This tells the agent that it should look into its credential store first and _only_ prompt the user for the password if the agent does not know one.
> 2. If `new_secrets` is `true`, the secret agent instead gets told `NM_SECRET_AGENT_GET_SECRETS_FLAG_REQUEST_NEW`. This means that the agent _must_ request a new password from the user.
> 
> If this code and the secret agent are working correctly, NetworkManager should attempt to reconnect to the network three times with the same credentials _before_ asking the user for a new password. I'm personally not sure what our agent is, but it's probably indicator-network. However, I would ask you to check and make sure that this behavior is working correctly. You can check this by using `dbus-monitor` to read the communication between NetworkManager and its clients, or you could add debugging prints in the code to check the values of `new_secrets` and `flags`. We're interesting in answering only one question: "Is our secret agent really being asked to prompt for the password, or is it mishandling the request?"
> 
> If the secret agent is really being told `NM_SECRET_AGENT_GET_SECRETS_FLAG_REQUEST_NEW`, then this patch is indeed one option to resolve the problem. However I'd be worried why `new_secrets` is being set right away, maybe that is what should really change.
> 
> Please don't take my comments as an attack on your work. However, I'm very concerned with the overall stability and security of Ubuntu Touch, and changing code inside network-manager in our downstream fork could very well compromise one or both of those. We need to be really careful with it. Upstream designed it a certain way for a reason, and we should be careful to understand why the walls are in place before we try to take them down.

Please skip the reasoning why you ask, and whether I should feel *attacked* by comments, etc... I'm used to work in upstreaming processes, I prefer to get straight to the point, and move forward.
 
> In that case, I would recommend opening a dialog with the upstream network-manager project. Hop in to #nm on Libera Chat IRC to ask about the design of this code path instead of assuming that it is wrong. See if upstream would take this patch as it is (the code hasn't changed much in the past 6 years). If not, we should probably be redesigning it too.

That is no problem, I'll get that up.

Thanks!

**Comment by UniversalSuperBox** [comment]
It indeed seems like we have a hole in our secret agent: https://github.com/ubports/indicator-network/blob/5899e1f83ff7dd3f388d295425fac06ede1e2cba/src/agent/SecretAgent.cpp#L229

I think this is where the fix should be targeted. If this replied correctly to `NM_SECRET_AGENT_GET_SECRETS_FLAG_ALLOW_INTERACTION` without `NM_SECRET_AGENT_GET_SECRETS_FLAG_REQUEST_NEW`, this problem would only crop up on the worst of networks that can fail authentication three times in a row.

File: debian/changelog [hunk 1]
```diff
@@ -1,3 +1,10 @@
+network-manager (1.2.2-1ubports4) xenial; urgency=medium
+
+  [Jos� Pekkarinen]
```
Comment (UniversalSuperBox) [inline debian/changelog]: Please use the ascii `é` instead of this `�` character that doesn't seem to exist on my system

File: debian/changelog [hunk 2]
```diff
@@ -1,3 +1,10 @@
...
+  * Remove connection attempts to unconfigured wifi
```
Comment (UniversalSuperBox) [inline debian/changelog]: Please fix this message to state what you're actually fixing.

**Review by UniversalSuperBox — CHANGES_REQUESTED** [review]

File: debian/changelog (hunk 2 above)
Comment (Pekkari) [reply debian/changelog]: I modified this line now as well as the commit message. Thanks!

File: debian/changelog (hunk 1 above)
Comment (Pekkari) [reply debian/changelog]: No utf-8? hu....

Comment (UniversalSuperBox) [reply debian/changelog]: I'm not sure which character you are using here, but it really does not exist on my system. None of Firefox, my terminal, or VS Code can print `�`.

**Review by UniversalSuperBox — APPROVED** [review]
After a lengthy review process, we've determined that the code that this calls has changed upstream in https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/commit/0eb76175047057fc466814632f25bfaf10566c37, perhaps accidentally. If I'm reading it right, this upstream change should prevent re-requesting secrets in the same situations that they are requested here.

Other upstream changes may also be a fix for this issue, with the change I've found being only a partial solution. However, we will be getting there soon. So we'll try this out while we're still on this older version of NM and see how it changes behavior.

**Comment by UniversalSuperBox** [comment]
Merged as https://github.com/ubports/network-manager-packaging/commit/d634c664617a4a4ecf91692c233f03e81e932671

**Comment by ivoxavier** [comment]
On my all devices (FP2,N5,6T) this is fixed for OTA-19 :)

**Pull request closed without merging**